ts_src/silentpayment.ts

```diff
@@ -0,0 +1,282 @@
+  'hex',
+);
+
+class SilentPaymentImpl implements SilentPayment {
+  constructor(private ecc: TinySecp256k1Interface) {}
+
+  /**
+   * Compute scriptPubKey used to send funds to a silent payment address
+   * @param inputs list of ALL outpoints of the transaction sending to the silent payment address
+   * @param inputPrivateKey private key owning the spent outpoint. Sum of all private keys if multiple inputs
+   * @param silentPaymentAddress target of the scriptPubKey
+   * @param index index of the silent payment address. Prevent address reuse if multiple silent addresses are in the same transaction.
+   * @returns the output scriptPubKey belonging to the silent payment address
+   */
+  makeScriptPubKey(
+    inputs: Outpoint[],
+    inputPrivateKey: Buffer,
+    silentPaymentAddress: string,
+    index = 0,
+  ): Buffer {
+    const inputsHash = hashOutpoints(inputs);
+    const addr = SilentPaymentAddress.decode(silentPaymentAddress);
+
+    const sharedSecret = this.makeSharedSecret(
+      inputsHash,
+      addr.scanPublicKey,
+      inputPrivateKey,
+    );
+
+    const outputPublicKey = this.makePublicKey(
+      addr.spendPublicKey,
+      index,
+      sharedSecret,
+    );
+
+    return Buffer.concat([SEGWIT_V1_SCRIPT_PREFIX, outputPublicKey.slice(1)]);
+  }
+
+  /**
+   * Check if a scriptPubKey belongs to a silent payment address
+   * @param scriptPubKey scriptPubKey to check
+   * @param inputs list of ALL outpoints of the transaction sending to the silent payment address
+   * @param inputPublicKey public key owning the spent outpoint. Sum of all public keys if multiple inputs
+   * @param scanSecretKey *scan* secret key of the silent payment address
+   * @param spendPublicKey *spend* public key of the silent payment address
+   * @param index index of the silent payment address.
+   */
+  isMine(
+    scriptPubKey: Buffer,
+    inputs: Outpoint[],
+    inputPublicKey: Buffer,
+    scanSecretKey: Buffer,
+    spendPublicKey: Buffer,
+    index = 0,
+  ): boolean {
+    const inputsHash = hashOutpoints(inputs);
+
+    const sharedSecret = this.makeSharedSecret(
+      inputsHash,
+      inputPublicKey,
+      scanSecretKey,
+    );
+
+    const outputPublicKey = this.makePublicKey(
+      spendPublicKey,
+      index,
+      sharedSecret,
+    );
+
+    return (
+      Buffer.compare(
+        scriptPubKey.slice(SEGWIT_V1_SCRIPT_PREFIX.length),
+        outputPublicKey.slice(1),
+      ) === 0
+    );
+  }
+
+  /**
+   * Compute the secret key used to spend an output locked by a silent address script.
+   * @param inputs outpoints of the transaction sending to the silent payment address
+   * @param inputPublicKey public key owning the spent outpoint in the tx (may be sum of public keys)
+   * @param spendSecretKey private key of the silent payment address
+   * @param index index of the silent payment address in the transaction, default to 0
+   * @returns 32 bytes key
+   */
+  makeSigningKey(
```

Comment:
We don't really need this "wrapper" function, I think it's better to export the building blocks (`makeSharedSecret`, `makeSecretKey`, `makePublicKey`) to give extreme flexibility to the consumer.